vmm/src/config.rs

```diff
@@ -2135,17 +2153,29 @@ pub struct RestoreConfig {
     pub source_url: PathBuf,
     #[serde(default)]
     pub prefault: bool,
+    #[serde(default)]
+    pub net_ids: Option<Vec<String>>,
+    #[serde(default)]
+    pub net_fds: Option<Vec<i32>>,
```

Comment:
I would work on what the `RestoreConfig` should look like and then work from there.

```rust
net_fds: Option<Vec<(String, i32)>>
```

But what about multiqueue devices - they have an fd for each queue pair so maybe it should be:

```rust
net_fds: Option<Vec<(String, Vec<i32>)>>
```

We can figure out how to make `OptionParser` handle it once we've decided on the struct.

Reply:
@rbradford  The info of how many FDs a net device (multiqueue) expects is already available from Snapshot's vmconfig. So, ``` net_fds: Option<Vec<i32>> ``` can be sequentially read as in the order of ids from `net_ids` and assigned each net id (dev).

@likebreath This will also eliminate the need for FDs count paired to a net dev. 

For example, for a VmConfig with 2 net devs - net1 with 2 queue pairs and net2 with 4 queue pairs,
the effective cmdline will look like 
``` --restore source_url=file:///foo/bar,net_ids=[net1,net2],net_fds=[23, 24, 27, 28, 29, 30]```.
net1 is assigned 23, 24 and net2 is assigned 27,28,29,30.

Reply:
I'm sorry to disagree - I think that's ugly and a hack and easy to make mistakes with - we support grouping things together in our command line parser so there is no excuse not use it.

Reply:
@rbradford The "sequential read of fds as in the order of net ids" was agreed on https://github.com/cloud-hypervisor/cloud-hypervisor/issues/6286 . The main reason for this approach was because all net devices' new FDs need to be passed via SCM_RIGHTS. 
The only additional change I made through this PR was to move  `[<net_id, fd_count>,...]` to just `[<net_id>,...]`. This was because the fd_count does not add any explicit value, as it is already available through the snapshot config.

Reply:
Now I've seen the implementation I can see that the internal `RestoreConfig` should be:

```
net_fds: Option<Vec<(String, Vec<i32>)>>
```

And that the CLI should take pairs of ID and a vector.

Sorry for the churn but sometimes these things only become apparent when they become more concrete.

Reply:
What i'm asking for is for you to "zip" i.e. pair up the real FDs with the ID when you construct the `RestoreConfig` struct in the API handler - rather than as including them as separate vectors. I know that the JSON won't contain any FDs (i.e. it would be an empty vector, or if using an `Option` type not included at all) but that is not a problem.

Reply:
Yeah, I thought about that. In order to pair up the real FDs with IDs (from the stream of FDs via SCM_RIGHTS), API handler need to be aware of how many FDs each ID should be associated with. This fd_count again either has to be passed via JSON which breaks the RestoreConfig data structure. Or the VM's snapshot config has to be read which I feel is not right to do in API handler. Do you have any inputs on how to handle this?

Reply:
You raise a really good point - although it would be duplicating information. I think it would strongly prefer sending the number of fds as part of the JSON protocol turning the struct into something like:

```
net_fds: Option<Vec<(String, i32, Option<Vec<i32>>)>>
```

To make the JSON ecoding nicer and to allow reuse for e.g. live migration

```rust
struct RestoredNetConfig {
    id: String,
    num_fds: usize,
    fds: Option<Vec<i32>>
}
```
and then placing it in array

```rust
struct RestoreConfig {
...
net_fds: Option<Vec<RestoredNetConfig>>
...
}

Reply:
Thanks @rbradford . This looks good to me.

I had dig at how to go about parsing. Tuple in parser supports only paris i.e, (S, T). An examlpe cli is ``` --cpus boot=3,affinity=[0@[2,3],1@[0,1]]```. But we will have a tuple of triplets for `net_fds`. So I made some rough changes here at https://github.com/cloud-hypervisor/cloud-hypervisor/commit/08ca5e0d1363cc9fd867445b0ffcfff25da23131 
It basically includes following changes:
- Support usage of '(' paranthesis. (changes at `split_commas` fn)
- Add ``` impl TupleValue for (usize, Vec<u64>) ```. This is one way to achieve triplet.
- Convert parsed arg to ```<Tuple<String, (usize, Vec<u64>)>>``` and fill RestoredNetconfig
- Effective cmd line will look like ```net_fds=[ net0@(2, [3,4]), net1@(1, [5]) ]```

One case that I still need to figure out is when`fds` has to be empty vector. Parser fails to read/convert when I give ```net_fds=[ net0@(2, []) ]``` 

Let me know your thoughts on this.

Reply:
> Thanks @rbradford . This looks good to me.
> 
> I had dig at how to go about parsing. Tuple in parser supports only pairs i.e, (S, T). An examlpe cli is ` --cpus boot=3,affinity=[0@[2,3],1@[0,1]]`. But we will have a Vec of triplets for `net_fds`. So I made some rough changes here at [08ca5e0](https://github.com/cloud-hypervisor/cloud-hypervisor/commit/08ca5e0d1363cc9fd867445b0ffcfff25da23131) It basically includes following changes:
> 
>     * Support usage of '(' paranthesis. (changes at `split_commas` fn)
> 
>     * Add `impl TupleValue for (usize, Vec<u64>)`. This is one way to achieve triplet.
> 
>     * Convert parsed arg to `<Tuple<String, (usize, Vec<u64>)>>` and fill RestoredNetconfig
> 
>     * Effective cmd line will look like `net_fds=[ net0@(2, [3,4]), net1@(1, [5]) ]`
> 
> 
> One case that I still need to figure out is when`fds` has to be empty vector. Parser fails to read/convert when I give `net_fds=[ net0@(2, []) ]`
> 
> Let me know your thoughts on this.

In this case I think we should have have the CLI interface deviate slightly from the HTTP API interface - we can work out  the number of FDs by what the user has provided. So I would go with

```
--restore source_url=file:///foo/bar, net_fds=[net1@[23, 24], net2@[27, 28, 29, 30]]
```

This would turn into a struct on the client:
```rust
net_fds = [ { id: "net1", num_fds: 2, fds: None }, { id:"net2", num_fds: 4, fds: None}]
```

But once received by the server and the fds used from SCM_RIGHTS:

```rust
net_fds = [ { id: "net1", num_fds: 2, fds: Some([123,124]) }, { id:"net2", num_fds: 4, fds: Some([127,128,129,130]) }]
```

(I incremented by 100 to represent that these are FDs in the CH process now)

Reply:
> Just to call out that `cloud-hypervisor` does not support passing FDs via SCM_RIGHTS. It parses cli to `RestoreConfig` and sends the request as is.

I actually regret that we have this alternative code path of being able to restore using the `cloud-hypervisor` binary directly - it predates us having the `ch-remote` tool and makes everything more complicated.